### ISSUE TITLE
Connection Problem

### ISSUE THREAD

**Issue by dwelltime** [beroset/pico-puller#2]
hello beroset

thanks for your great tool in first place.

i compiled everything successfully ( i think) flashed the uf2 file to the pico and can access via terminal.
when i type for example version it shows me only version "".

my pinout how i connected to the m16c works flawless with e8a debugger. and shows me for example ver 2.77.
i saw under windows that i have an additional unknown "RESET" device.im not sure if its normal.

is there any chance to debug what is wrong with my connection? or maybe some options than i can change so i can connect successfully?

greetings

**Comment by beroset** [comment]
Do you get the menu title?  It should say "pico-puller Main Menu:" and then list the menu.  If that works, but `version` does not, then there are two possibilities: either the target device does not implement the Renesas bootloader (and so ignores the command) or there is a problem with the connection between your Pico and the M16C.  Verify carefully that the pin connections shown in the README file are *exactly* the ones you are using.  Also, make sure you know whether it's a square M16C package or a rectangular one -- the pin numbers are different.

**Comment by speters** [comment]
Thx for this software!

I do not know if it is the same behaviour, but I am also not able to get a responsive terminal connection.

The "help" dialog is displayed during connecting.
A "help" (with \n line ending) shows the help. "quit" works while terminal is responsive (which means after booting or after a "help").

An unknown command like "h" shows "> That is not a valid choice.". Then terminal is unresponsive.

Once terminal is unresponsive, rp2040 has to be reset (unplugging USB).

A command which does not [block](https://github.com/beroset/pico-puller/blob/9393e8495152c387c9e96e8f2edd56a417237ef7/src/M16C.cpp#L78) for pio fifo read/write like "set_speed 1e6" does not work, terminal does not [respond](https://github.com/beroset/pico-puller/blob/9393e8495152c387c9e96e8f2edd56a417237ef7/src/main.cpp#L227C31-L227C53).

Any other command with blocking pio fifo ops also does not work, but this could be due to my setup with an [M3062GF8NGP](https://www.renesas.com/us/en/document/dst/m3062gf8nfpgp-datasheet) on a somewhat unknown board (Navman Repeat 3100s).

Tried with current main HEAD ([9393e84](https://github.com/beroset/pico-puller/commit/9393e8495152c387c9e96e8f2edd56a417237ef7)), and also with the uf2 binary from the 1.1 release (which seems to have different pins defined).
PICO SDK version 1.5.1

Same on RP2040 B1 and RP2040 B2 silicon.

**Comment by beroset** [comment]
@speters: Thanks for your comment.  Are you also using a Windows host to connect to the `pico`?  I am wondering if the the line endings might be part of the issue.  My test and development platform is Linux where the line ending is `\n` versus Windows where the line ending is typically `\r\n`.  I see that you've said your line ending is set to `\n` but, depending on the terminal program used, sometimes on Windows that still gets translated to `\r\n`.  I don't have a Windows machine to try this on, but I will experiment with line endings to see if there's something that I could improve here.

As for pinout, the 1.1 version should *definitely* be used since the software and documentation were both changed to match.

**Comment by speters** [comment]
Thx for your reply.

I am testing on Linux with "minicom" terminal, line endings set to "\n".

Tag 1.1 is commit [5343592](https://github.com/beroset/pico-puller/commit/53435928bf4716ca45020791b929b79abf7e9e40) which has different pin settings (clkpin 2, resetpin 6).
With this 1.1 firmware, the terminal behaves similar:
 * help displayed on connect
 * "`help`" displays help
 * unknown command (e.g. "sdsad") is responded with "`That is not a valid choice`", then unresponsive
 * non-blocking command "`set_speed 1.0e6`" does nothing but making terminal unresponsive
 * "`version`" commands replies with "` Version: ""`"
 * other commands make terminal unresponsive

My connections to the M16C are as written in current README.md. (clkpin 18, resetpin 22), I used pico-puller.uf2 compiled from HEAD revision.
 * difference to 1.1: "`version`" command makes terminal unresponsive (no output)

Regards from Hamburg, Germany,
Sönke

**Comment by dwelltime** [comment]
hi beroset hi speters

Do you get the menu title?  **yes i do as i wrote**
It should say "pico-puller Main Menu:" and then list the menu. **yes it does**

If that works, but version does not, then there are two possibilities: either the target device does not implement the Renesas bootloader (and so ignores the command) **i can connect with other tools bootloader is VER.2.77**


or there is a problem with the connection between your Pico and the M16C. **i must say i have exactly the same behavior as speters wrote**

**Comment by beroset** [comment]
OK, I've done two things that may be helpful here.  First, I created a 1.2 release.  No code was changed except for the version number, but I thought it would be useful to have the latest version available for people who don't want to build it themselves.  

Second, I experimented with `minicom` on Linux and see problems very similar to what you describe.  The problem is that `minicom` sends a CR instead of LF.  It's only the LF that the pico-puller software understands as the end of a line.  The way to avoid that problem is to tell `minicom` to use LF instead.  Go into "Screen and Keyboard" setup, choose "O - Character conversion" and change character 13 (CR) to emit 10 (LF).  That should take care of part of the problem.  If you're using some other terminal software, either do something similar or perhaps tell your software to use the "DOS" line ending pair (CR/LF).

The other thing is that you probably need to issue a "reset 0" to release the reset line of the CPU.  Otherwise it will be held in reset and nothing will work.  Look at the `dump.cmd` file in the `scripts` directory for a typical sequence required to dump memory. 

Lastly, using the `expect` scripts provided with the project works much better than a terminal emulator.  Try them (the `pico-puller` command) if you're on Linux.

Please let me know your results.  If one or more of these suggestions work for you, I can see about adding to the documentation to make things more clear.  Thanks for your help!

**Comment by speters** [comment]
I am now starting to think that I have run into a hardware problem here. I think I will have to remove the µC from the PCB and hook it up with a minimal configuration. I'll report back if I am successful.
Thx for your kind support!

**Comment by dwelltime** [comment]
thanks beroset

i have the mcu hooked up on a breakout board.

**Comment by beroset** [comment]
@speters @dwelltime Have you been successful?

**Comment by speters** [comment]
thx for asking, @beroset.
Unfortunately the device I worked on got damaged under my investigations.

I did another try with a M3062GF8NGP of questionable provenance on a break-out board. As I had no XTAL at hand, I used the RP2040 to generate an ~8MHz clock. For this, I changed current HEAD revision of the pico-puller to use the pinout of v1 in order to have gpio 21 free for use by CLOCK GPOUT0. Clock was fed on XIN of the M16C with XOUT left unconnected as the datasheet suggests.
With this setup, the pico-puller was reactive, but would only read back 0xff. Tried with various clock speeds, but no luck.

I ended the experiment, as the original DUT no longer was in working condition.

Thanks four software and especially for your very kind support! I hope I didnt't bother too much with my problems.

From my POV this issue could be closed, as the pico-puller was not causal for the problem.
